File: day5-handson/python/src/user_service.py
# user_service.py
#
# This is the UNIT UNDER TEST for Task 3: Fakes.
#
# UserService orchestrates user management operations.
# It depends on a repository to persist data.
#
# The repository is injected via __init__ — DEPENDENCY INJECTION.
# In production: UserService(RealUserRepository())
# In tests:      UserService(FakeUserRepository())
#
# Business rules:
#   - Users must have an id, name, and email.
#   - Email must contain "@".
#   - Duplicate ids are rejected.
#   - list_all() returns users sorted by name (A → Z).
#   - remove_user() returns False if the user didn't exist.


class UserService:
    """
    Manages user registration, lookup, listing, and removal.

    Args:
        repository: Any object implementing save, find_by_id,
                    find_all, and delete_by_id.
                    Injected so tests can pass a FakeUserRepository.
    """

    def __init__(self, repository):
        self.repository = repository

    def register_user(self, user: dict) -> dict:
        """
        Registers a new user after validating the input.

        Args:
            user: dict with keys 'id', 'name', 'email'

        Returns:
            dict with 'success' bool and optional 'reason' string.
        """
        # Validate required fields
        if not user or not all(k in user for k in ('id', 'name', 'email')):
            return {'success': False, 'reason': 'Missing required fields: id, name, email'}

        # Validate email format
        if '@' not in user['email']:
            return {'success': False, 'reason': 'Invalid email address'}

        # Check for duplicate id
        existing = self.repository.find_by_id(user['id'])
        if existing:
            return {'success': False, 'reason': f"User with id \"{user['id']}\" already exists"}

        self.repository.save(user)
        return {'success': True}

    def get_user(self, user_id: str) -> dict | None:
        """
        Retrieves a single user by id.

        Returns the user dict, or None if not found.
        """
        return self.repository.find_by_id(user_id)

    def list_all(self) -> list[dict]:
        """
        Returns all users sorted alphabetically by name (A → Z).
        """
        users = self.repository.find_all()
        return sorted(users, key=lambda u: u['name'])

    def remove_user(self, user_id: str) -> bool:
        """
        Removes a user. Returns False if the user didn't exist.
        """
        return self.repository.delete_by_id(user_id)

File: week-1/day5-handson/js/src/userService.js
/**
 * userService.js
 *
 * This is the UNIT UNDER TEST for Task 3: Fakes.
 *
 * UserService orchestrates user management operations.
 * It depends on a repository to persist data.
 *
 * The repository is injected via the constructor — this is
 * DEPENDENCY INJECTION. In production we pass a RealUserRepository.
 * In tests we pass a FakeUserRepository.
 *
 * Business rules:
 *   - Users must have an id, name, and email.
 *   - Email must contain "@".
 *   - Duplicate ids are rejected.
 *   - listAll() returns users sorted by name (A → Z).
 *   - removeUser() returns false if the user didn't exist.
 */

class UserService {
  /**
   * @param {object} repository - Any object with save / findById /
   *                              findAll / deleteById methods.
   *                              Injected so tests can pass a Fake.
   */
  constructor(repository) {
    this.repository = repository;
  }

  /**
   * Registers a new user after validating the input.
   *
   * @param {{ id: string, name: string, email: string }} user
   * @returns {Promise<{ success: boolean, reason?: string }>}
   */
  async registerUser(user) {
    // Validate required fields
    if (!user?.id || !user?.name || !user?.email) {
      return { success: false, reason: 'Missing required fields: id, name, email' };
    }

    // Validate email format
    if (!user.email.includes('@')) {
      return { success: false, reason: 'Invalid email address' };
    }

    // Check for duplicate
    const existing = await this.repository.findById(user.id);
    if (existing) {
      return { success: false, reason: `User with id "${user.id}" already exists` };
    }

    await this.repository.save(user);
    return { success: true };
  }

  /**
   * Retrieves a single user by id.
   *
   * @param {string} id
   * @returns {Promise<object|null>} The user, or null if not found.
   */
  async getUser(id) {
    return this.repository.findById(id);
  }

  /**
   * Returns all users sorted alphabetically by name.
   *
   * @returns {Promise<object[]>}
   */
  async listAll() {
    const users = await this.repository.findAll();
    return users.sort((a, b) => a.name.localeCompare(b.name));
  }

  /**
   * Removes a user. Returns false if the user didn't exist.
   *
   * @param {string} id
   * @returns {Promise<boolean>}
   */
  async removeUser(id) {
    return this.repository.deleteById(id);
  }
}

module.exports = { UserService };
